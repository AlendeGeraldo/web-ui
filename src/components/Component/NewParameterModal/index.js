/**
 * Component responsible for:
 * - Structuring the advanced new component parameter layout
 * - Add parameter
 */
import React from 'react';
import { connect } from 'react-redux';
import PropTypes from 'prop-types';
import { Form, Input, message, Modal, Select, Switch } from 'antd';
import { addParam } from '../../../store/actions/componentActions';

const { Option } = Select;
const { TextArea } = Input;

const NewParameterModal = (props) => {
  const { details, form, visible, onCancel, onAddParam } = props;
  const { uuid, parameters } = details;
  const { getFieldDecorator, getFieldsError } = form;

  /**
   * Function used to check if form has errors
   * @param {string[]} fieldsError
   */
  const hasErrors = (fieldsError) => {
    return Object.keys(fieldsError).some((field) => fieldsError[field]);
  };

  /**
   * Function to handle modal cancel
   */
  const handleCancel = () => {
    form.resetFields();
    onCancel();
  };

  /**
   * Function to handle form submit
   * @param {Event} e
   */
  const handleSubmit = (e) => {
    e.preventDefault();
    form.validateFields((err, values) => {
      if (!err) {
        const newParameter = {
          name: values.name,
          type: values.type,
          required: values.required,
          default: values.defaultValue,
          details: values.details,
        };

        onAddParam(uuid, parameters, newParameter).then((response) => {
          if (response) {
            message.success(
              `Par창metro ${newParameter.name} adicionado com sucesso`
            );
            handleCancel();
          }
        });
      }
    });
  };

  return (
    <Modal
      visible={visible}
      title='Novo Par창metro'
      okText='Criar'
      cancelText='Cancelar'
      onCancel={handleCancel}
      onOk={handleSubmit}
      okButtonProps={{ disabled: hasErrors(getFieldsError()) }}
    >
      <Form layout='vertical'>
        <Form.Item label='Nome: '>
          {getFieldDecorator('name', {
            rules: [{ required: true, message: ' ' }],
          })(<Input />)}
        </Form.Item>
        <Form.Item label='Tipo: '>
          {getFieldDecorator('type', {
            rules: [{ required: true, message: ' ' }],
          })(
            <Select>
              <Option value='float'>Float</Option>
              <Option value='int'>Int</Option>
              <Option value='string'>String</Option>
            </Select>
          )}
        </Form.Item>
        <Form.Item label='Valor padr찾o: '>
          {getFieldDecorator('defaultValue')(<Input />)}
        </Form.Item>
        <Form.Item label='Detalhes: '>
          {getFieldDecorator('details')(<TextArea />)}
        </Form.Item>
        <Form.Item label='Obrigat처rio: '>
          {getFieldDecorator('required', {
            valuePropName: 'checked',
            initialValue: true,
          })(<Switch />)}
        </Form.Item>
      </Form>
    </Modal>
  );
};

NewParameterModal.propTypes = {
  visible: PropTypes.bool.isRequired,
  onCancel: PropTypes.func.isRequired,
  form: PropTypes.shape({
    getFieldDecorator: PropTypes.func,
    getFieldsError: PropTypes.func,
  }).isRequired,
};

const mapStateToProps = (state) => {
  return {
    details: state.component.details,
  };
};

const mapDispatchToProps = (dispatch) => {
  return {
    onAddParam: (id, parameters, newParameter) => {
      return dispatch(addParam(id, parameters, newParameter));
    },
  };
};

export default connect(
  mapStateToProps,
  mapDispatchToProps
)(Form.create({ name: 'new-parameter-modal' })(NewParameterModal));
