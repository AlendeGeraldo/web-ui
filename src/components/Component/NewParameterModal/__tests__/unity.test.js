import React from 'react';
import { Provider } from 'react-redux';
import thunk from 'redux-thunk';
import { shallow, mount } from 'enzyme';
import configureMockStore from 'redux-mock-store';
import { Input, Modal, Select } from 'antd';
import NewParameterModal from '..';

const component = {
  uuid: '1',
  createdAt: '2019-11-05T17:16:53.000Z',
  name: 'AutoML',
  parameters: [
    { name: 'b', type: 'int', required: true },
    { name: 'a', type: 'int', required: false },
  ],
};

const initialState = {
  component: {
    details: component,
  },
};

const middlewares = [thunk];
const mockStore = configureMockStore(middlewares);
const store = mockStore(initialState);
const onCancel = () => {};

describe('NewParameterModal component', () => {
  afterEach(() => {
    jest.clearAllMocks();
  });

  it('is expected render without crashing', () => {
    shallow(
      <Provider store={store}>
        <NewParameterModal visible onCancel={onCancel} />
      </Provider>
    );
  });

  it('is expected render html correctly', () => {
    const wrapper = shallow(
      <Provider store={store}>
        <NewParameterModal visible onCancel={onCancel} />
      </Provider>
    );
    expect(wrapper).toMatchSnapshot();
  });

  it('should render self and subcomponents', () => {
    const wrapper = mount(
      <Provider store={store}>
        <NewParameterModal visible onCancel={onCancel} />
      </Provider>
    );

    expect(wrapper.find('Form(NewParameterModal)').exists()).toBeTruthy();
    expect(wrapper.find(NewParameterModal).exists()).toBeTruthy();
    expect(wrapper.find(Modal).exists()).toBeTruthy();
  });

  it('onOk should be called', () => {
    // work around to ignore console log antd
    jest.spyOn(console, 'warn').mockImplementation(() => {});
    jest.spyOn(console, 'error').mockImplementation(() => {});

    const wrapper = mount(
      <Provider store={store}>
        <NewParameterModal visible onCancel={onCancel} />
      </Provider>
    );

    const event = { preventDefault: () => {} };
    const modal = wrapper.find(Modal).instance();
    const spyHandleOk = jest.spyOn(modal, 'handleOk');
    modal.handleOk(event);
    expect(spyHandleOk).toHaveBeenCalled();
  });

  it('onCancel should be called', () => {
    const wrapper = mount(
      <Provider store={store}>
        <NewParameterModal visible onCancel={onCancel} />
      </Provider>
    );

    const modal = wrapper.find(Modal).instance();
    const spyHandleCancel = jest.spyOn(modal, 'handleCancel');
    modal.handleCancel();
    expect(spyHandleCancel).toHaveBeenCalled();
  });

  it('handleSubmit', () => {
    // mock dispatch response
    store.dispatch = jest.fn().mockImplementation(() => {
      return Promise.resolve(true);
    });

    const wrapper = mount(
      <Provider store={store}>
        <NewParameterModal visible onCancel={onCancel} />
      </Provider>
    );

    // change input value
    wrapper
      .find(Input)
      .at(0)
      .simulate('change', {
        target: { value: 'name' },
      });

    // change select value
    wrapper.find(Select).simulate('click');
    wrapper
      .find('MenuItem')
      .at(0)
      .simulate('click');

    const event = { preventDefault: () => {} };
    const modal = wrapper.find(Modal).instance();
    modal.handleOk(event);
    expect(store.dispatch).toHaveBeenCalledTimes(1);
  });
});
