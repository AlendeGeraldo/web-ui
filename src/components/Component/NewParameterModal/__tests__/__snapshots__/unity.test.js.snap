// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NewParameterModal component is expected render html correctly 1`] = `
<Modal
  cancelText="Cancelar"
  confirmLoading={false}
  maskTransitionName="fade"
  okButtonProps={
    Object {
      "disabled": false,
    }
  }
  okText="Criar"
  okType="primary"
  onCancel={[Function]}
  onOk={[Function]}
  title="Novo Parâmetro"
  transitionName="zoom"
  visible={true}
  width={520}
>
  <Form
    colon={true}
    hideRequiredMark={false}
    layout="vertical"
    onSubmit={[Function]}
  >
    <FormItem
      hasFeedback={false}
      label="Nome: "
    >
      <Input
        data-__field={
          Object {
            "name": "name",
          }
        }
        data-__meta={
          Object {
            "name": "name",
            "originalProps": Object {
              "type": "text",
            },
            "ref": null,
            "rules": Array [
              Object {
                "message": " ",
                "required": true,
              },
            ],
            "trigger": "onChange",
            "validate": Array [
              Object {
                "rules": Array [
                  Object {
                    "message": " ",
                    "required": true,
                  },
                ],
                "trigger": Array [
                  "onChange",
                ],
              },
            ],
            "valuePropName": "value",
          }
        }
        id="new-parameter-modal_name"
        onChange={[Function]}
        type="text"
      />
    </FormItem>
    <FormItem
      hasFeedback={false}
      label="Tipo: "
    >
      <Select
        choiceTransitionName="zoom"
        data-__field={
          Object {
            "name": "type",
          }
        }
        data-__meta={
          Object {
            "name": "type",
            "originalProps": Object {
              "children": Array [
                <Option
                  value="float"
                >
                  Float
                </Option>,
                <Option
                  value="int"
                >
                  Int
                </Option>,
                <Option
                  value="string"
                >
                  String
                </Option>,
              ],
              "choiceTransitionName": "zoom",
              "showSearch": false,
              "transitionName": "slide-up",
            },
            "ref": null,
            "rules": Array [
              Object {
                "message": " ",
                "required": true,
              },
            ],
            "trigger": "onChange",
            "validate": Array [
              Object {
                "rules": Array [
                  Object {
                    "message": " ",
                    "required": true,
                  },
                ],
                "trigger": Array [
                  "onChange",
                ],
              },
            ],
            "valuePropName": "value",
          }
        }
        id="new-parameter-modal_type"
        onChange={[Function]}
        showSearch={false}
        transitionName="slide-up"
      >
        <Option
          value="float"
        >
          Float
        </Option>
        <Option
          value="int"
        >
          Int
        </Option>
        <Option
          value="string"
        >
          String
        </Option>
      </Select>
    </FormItem>
    <FormItem
      hasFeedback={false}
      label="Valor padrão: "
    >
      <Input
        data-__field={
          Object {
            "name": "defaultValue",
          }
        }
        data-__meta={
          Object {
            "name": "defaultValue",
            "originalProps": Object {
              "type": "text",
            },
            "ref": null,
            "trigger": "onChange",
            "validate": Array [],
            "valuePropName": "value",
          }
        }
        id="new-parameter-modal_defaultValue"
        onChange={[Function]}
        type="text"
      />
    </FormItem>
    <FormItem
      hasFeedback={false}
      label="Detalhes: "
    >
      <TextArea
        data-__field={
          Object {
            "name": "details",
          }
        }
        data-__meta={
          Object {
            "name": "details",
            "originalProps": Object {},
            "ref": null,
            "trigger": "onChange",
            "validate": Array [],
            "valuePropName": "value",
          }
        }
        id="new-parameter-modal_details"
        onChange={[Function]}
      />
    </FormItem>
    <FormItem
      hasFeedback={false}
      label="Obrigatório: "
    >
      <Switch
        checked={true}
        data-__field={
          Object {
            "name": "required",
          }
        }
        data-__meta={
          Object {
            "initialValue": true,
            "name": "required",
            "originalProps": Object {},
            "ref": null,
            "trigger": "onChange",
            "validate": Array [],
            "valuePropName": "checked",
          }
        }
        id="new-parameter-modal_required"
        onChange={[Function]}
      />
    </FormItem>
  </Form>
</Modal>
`;
